src/foam/core/Validation.js

```diff
@@ -15,15 +15,151 @@
  * limitations under the License.
  */
 
+foam.CLASS({
+  package: 'foam.core',
+  name: 'ValidationPredicate',
+  properties: [
+    {
+      name: 'predicateFactory'
+    },
+    {
+      class: 'FObjectProperty',
+      of: 'foam.mlang.predicate.Predicate',
+      name: 'predicate',
+      expression: function(predicateFactory) {
+        return predicateFactory ?
+          predicateFactory(foam.mlang.ExpressionsSingleton.create()) :
+          null;
+      }
+    },
+    {
+      class: 'StringArray',
+      name: 'args'
+    },
+    {
+      class: 'Function',
+      name: 'jsFunc',
+      expression: function(predicate, jsErr) {
+        return function() {
+          if ( ! predicate.f(this) ) return jsErr(this);
+        };
+      }
+    },
+    {
+      class: 'String',
+      name: 'errorString'
+    },
+    {
+      class: 'Function',
+      name: 'jsErr',
+      expression: function(errorString) {
+        return function() { return errorString; };
+      }
+    }
+  ],
+  methods: [
+    function createErrorSlotFor(data) {
+      return this.ExpressionSlot.create({
+        args: this.args.map(a => data[a+'$']),
+        code: this.jsFunc.bind(data)
+      });
+    }
+  ]
+});
+
 foam.CLASS({
   package: 'foam.core',
   name: 'PropertyValidationRefinement',
   refines: 'foam.core.Property',
 
   properties: [
+    {
+      class: 'FObjectArray',
+      of: 'foam.core.ValidationPredicate',
+      name: 'validationPredicates'
+    },
+    {
+      name: 'validateObj',
+      expression: function(name, label, required, validationPredicates) {
+        if ( validationPredicates.length ) {
+          var args = foam.Array.unique(validationPredicates
+            .map(vp => vp.args)
+            .flat());
+          return [args, function() {
+            for ( var i = 0 ; i < validationPredicates.length ; i++ ) {
+              var vp = validationPredicates[i];
+              if ( vp.jsFunc.bind(this)() ) return vp.jsErr.bind(this)();
+            }
+            return null;
+          }];
+        }
+        return !required ? null : [[name],
+          function() {
+            return !this.hasOwnProperty(name) && (label + ' is required.');
+          }]
+      },
+    },
+  ]
+});
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'StringPropertyValidationRefinement',
+  refines: 'foam.core.String',
+  properties: [
+    'minLength',
+    'maxLength',
+    {
+      class: 'FObjectArray',
+      of: 'foam.core.ValidationPredicate',
+      name: 'validationPredicates',
+      factory: function() {
+        var self = this;
+        var a = [];
+        if ( foam.Number.isInstance(this.minLength) ) {
+          a.push({
+            args: [this.name],
+            predicateFactory: function(e) {
+              return e.REG_EXP(self, new RegExp('^.{'+this.minLength+',}$'));
```

Comment:
A RegExp for a minimum length feels like overkill.

Reply:
Can't disagree with that :)
What do you suggest instead? Do we have an mlang expression for length of a string or something? If not, should I create one or did you have something else in mind?

Reply:
mlang expression for string length would be useful I think

Reply:
On it

Reply:
Done